lib/mix/lib/mix/tasks/compile.all.ex

```diff
@@ -12,6 +12,13 @@ defmodule Mix.Tasks.Compile.All do
   @impl true
   def run(args) do
     Mix.Project.get!()
+
+    Mix.Lock.with_lock(Mix.Tasks.Compile.compile_lock_key(), fn ->
+      do_run(args)
+    end)
```

Comment:
Let's encapsulate this in `Mix.Task.Compiler.with_lock(fn -> ... end)` instead of passing `compile_lock_key` around. It can be `@doc false` for now but we will eventually want to make it public. :)

Reply:
Or even better, let's pass `config` as argument to the function so we fetch it less often?

Reply:
Or maybe it should be in `Mix.Project.with_build_lock`.